srml/support/src/dispatch.rs

```diff
@@ -808,40 +807,46 @@ macro_rules! __impl_outer_dispatch_metadata {
 		$( $module:ident::$call:ident, )*
 	) => {
 		impl $runtime {
-			pub fn outer_dispatch_metadata() -> $crate::dispatch::OuterDispatchMetadata {
-				$crate::dispatch::OuterDispatchMetadata {
-					name: $crate::dispatch::DecodeDifferent::Encode(stringify!($outer_name)),
-					calls: __impl_outer_dispatch_metadata!(@encode_calls 0; ; $( $module::$call, )*),
-				}
+			pub fn module_dispatch(mod_name: &'static str) -> Option<$crate::dispatch::OuterDispatchCall> {
+				__impl_outer_dispatch_metadata!(@filter mod_name; 0; ; $( $module::$call, )*);
+				None
 			}
 		}
 	};
-	(@encode_calls
+	(@filter
+		$mod_name:ident;
 		$index:expr;
-		$( $encoded_call:expr ),*;
+		$( $encoded_call:expr )*;
 		$module:ident::$call:ident,
 		$( $rest_module:ident::$rest:ident, )*
 	) => {
 		__impl_outer_dispatch_metadata!(
-			@encode_calls
+			@filter
+			$mod_name;
 			$index + 1;
-			$( $encoded_call, )*
-			$crate::dispatch::OuterDispatchCall {
-				name: $crate::dispatch::DecodeDifferent::Encode(stringify!($call)),
-				prefix: $crate::dispatch::DecodeDifferent::Encode(stringify!($module)),
-				index: $index,
-			};
+			$( $encoded_call )*
+			if $mod_name == stringify!($module) {
```

Comment:
Did you see my proposal to use https://crates.io/crates/paste? With `paste` you will be able to create one function per outer dispatch module. So, the creation of the metadata is still an almost static operation. 
I'm strongly on adding to much runtime code, especially using `ifs` for the generation.

Reply:
I can also create one function per module by putting them in a dedicated module (no need for 'paste'). I did not go this way because I think it makes a hard relation between macros. I will check if it is the case by panicking in the case where it does not find the module.

What is the biggest cost, multiple function definition or some ifs, ok functions will be inlined, but does the call with always the same static string parameter (single execution path) get some optim too?

While writing this I did test replacing the no module found by exit condition and it did panic.
I do not really see how I can generate a default function for module not include in impl_outer_event or impl_outer_dispatch.

Reply:
Ok I understand the need for 'paste' :) modules won't do it.

Reply:
Okay, perfect :)

Reply:
Ok it works for the metadata test case ,  but breaks over the actual substrate runtime (current branch : https://github.com/paritytech/substrate/compare/master...cheme:metadata3 ). 
I got errors like:
``` 
no associated item named `__module_events_consensus` found for type `Runtime` in the current scope
   --> node/runtime/src/lib.rs:225:1
    |
225 |    construct_runtime!(
    |   _^
    |  |_|
    | ||
226 | ||     pub enum Runtime with Log(InternalLog: DigestItem<Hash, SessionKey>) where
227 | ||         Block = Block,
228 | ||         NodeBlock = node_primitives::Block,
...   ||
248 | ||     }
249 | || );
    | ||  ^
    | ||__|
    | |___associated item `__module_events_consensus` not found for this
    |     associated item not found in `Runtime`
    |
    = help: did you mean `__module_events_contract`?
```
And indeed srml/contract got a call to decl_events when srml/consensus does not.

I am back to my first intuition... So if we want to pursue that way I may probably need to create a decl_no_event or adapt decl_event for allowing empty content and then it must be  that every module must call one of them.
I got some other errors more cryptic so I am not sure wether it is ok for outer dispatch.

Reply:
You probably just need to change what the `impl_runtime_metadata!` macro takes as input. For example we already have the case with `with Storage` and without that. All information you need are available in the `construct_runtime!` macro. This already knows if a module has Events, Storage, Calls, etc. `construct_runtime!` is also calling this macro, so you just need to change this call to give you more information.

Reply:
It just calls `store_metadata` when the module defines a storage and we know that, because the input parameters for the macro say it so.
If you have a module with storage, the macro call will look like:
```
impl_runtime_metadata! {
   BlaBla
      MyModule with Storage
}
```

Now in `construct_runtime!` you say which stuff a module provide, so it looks like:
```
construct_runtime! {
    BlaBlaBla
        MyModule: my_module::{Event, Storage, Call,...},
}
```

So we have already a list of all the stuff that is supported by a module. You just need to forward this information to the `impl_runtime_metadata` call.

Reply:
Makes sense (looks like the macro will grow a tiny bit).